services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/s3crt/RequestDataSupplierAdapter.java

```diff
@@ -0,0 +1,160 @@
+package software.amazon.awssdk.services.s3.internal.s3crt;
+
+import com.amazonaws.s3.RequestDataSupplier;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public class RequestDataSupplierAdapter implements RequestDataSupplier, Subscriber<ByteBuffer> {
+    private final Lock lock = new ReentrantLock();
+    private final Publisher<ByteBuffer> bodyPublisher;
+    private final Deque<Event> eventBuffer = new ArrayDeque<>();
+
+    private volatile Subscription subscription;
+    private volatile Condition eventAvailable = lock.newCondition();
+    private volatile boolean dataRequested = false;
+
+    public RequestDataSupplierAdapter(Publisher<ByteBuffer> bodyPublisher) {
+        this.bodyPublisher = bodyPublisher;
+        this.bodyPublisher.subscribe(this);
+    }
+
+    @Override
+    public boolean getRequestBytes(ByteBuffer bytes) {
```

Comment:
I've doing some testing against this updated interface, and CRT doesn't seem to treat the return value the way I'd expect, given this description.

 I would expect the following to work
 
```java
getRequestBytes(ByteBuffer bb) {
    return false;
}
```

which results in `getRequestBytes()` being called continuously, but returning false results in the request completing instead.

Reply:
Oops, the docs are referring to are here https://github.com/awslabs/aws-crt-java/blob/06593b80e3497a74a27b87a59a9f5c4b7c07cd56/s3-native-client/src/main/java/com/amazonaws/s3/RequestDataSupplier.java#L14-L17

Reply:
Discussing with CRT team about what we can do to support the use case above. For now, the implementation fills the buffer before returning